validator/enforce_labels.rego

```diff
@@ -35,35 +26,58 @@ deny[{
 	lib.get_constraint_params(constraint, params)
 	asset := input.asset
 
-	mandatory_labels := params.mandatory_labels[_]
-	resource_is_to_scan(asset)
-	not label_is_valid(mandatory_labels, asset)
+	mandatory_label := params.mandatory_labels[_]
+	label_value_pattern := mandatory_label[label_key]
+
+	tested_resource_types := [
+		"cloudresourcemanager.googleapis.com/Project",
+		"storage.googleapis.com/Bucket",
+		"compute.googleapis.com/Instance",
+		"compute.googleapis.com/Image",
+		"compute.googleapis.com/Disk",
+		"compute.googleapis.com/Snapshot",
+		"google.bigtable.Instance",
+		"sqladmin.googleapis.com/Instance",
+	]
+
+	non_standard_types := ["sqladmin.googleapis.com/Instance"]
 
-	message := sprintf("%v doesn't have a required label.", [asset.name])
+	resource_types_to_scan := lib.get_default(params, "resource_types_to_scan", tested_resource_types)
 
-	metadata := {"resource": asset.name, "mandatory_labels": mandatory_labels}
+	resource_is_to_scan(asset, resource_types_to_scan)
+
+	not label_is_valid(label_key, label_value_pattern, asset, non_standard_types)
+
+	message := sprintf("%v's label is in violation.", [asset.name])
+	metadata := {"resource": asset.name, "label_in_violation": label_key}
 }
 
-resource_is_to_scan(asset) {
+resource_is_to_scan(asset, resource_types_to_scan) {
 	resource_types_to_scan[_] == asset.asset_type
 }
 
-# generic label_is_valid for all resources
-label_is_valid(label, asset) {
-	resource := get_label(asset)
-	resource_labels := lib.get_default(resource, "labels", {})
-	labelValue := resource_labels[label]
+# check if label exists for all resources to scan
+label_is_valid(label_key, label_value_pattern, asset, non_standard_types) {
+	# retrieve the right values from asset
+	resource_labels := get_labels(asset, non_standard_types)
+
+	# test if label exists in asset
+	resource_labels[label_key]
+
+	# test if label value matches pattern passed as a parameter 
+re_match(	label_value_pattern, resource_labels[label_key])
 }
 
-# get_generic_label Object for standard resources
-get_generic_label(asset) = resource {
+# get_labels for standard resources
+get_labels(asset, non_standard_types) = resource_labels {
+	asset.asset_type != non_standard_types[_]
```

Comment:
is this right? we seem to be checking that asset_type != **any** one of the standard types here.

Reply:
I think it's correct, this test will return false if the asset type is non-standard which is what we want here: only standard types have the labels under "labels" in asset.resource.data.

Each non-standard type (or group of non-standard types) will have their own get_labels function to retrieve the labels in their custom location (like the sql instances in this version).

Reply:
oops. I meant to say checking that asset_type != any one of the non-standard types. It happens to work here because there's only 1 non-standard type. But if there's more than 1, it's not going to give the correct result. Example: https://play.openpolicyagent.org/p/AY0g0knZ2r

Reply:
Good catch! I did not think about this. I flipped the logic around to test exactly for standard types instead since this is what I was testing for anyway, it seems to work fine.